weave/trace_server/calls_query_builder.py

```diff
@@ -0,0 +1,726 @@
+    field: str
+
+    def as_sql(
+        self,
+        pb: ParamBuilder,
+        table_alias: str,
+        cast: typing.Optional[tsi_query.CastTo] = None,
+    ) -> str:
+        return clickhouse_cast(f"{table_alias}.{self.field}", cast)
+
+    def as_select_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        return f"{self.as_sql(pb, table_alias)} AS {self.field}"
+
+    def is_heavy(self) -> bool:
+        return False
+
+
+class CallsMergedAggField(CallsMergedField):
+    agg_fn: str
+
+    def as_sql(
+        self,
+        pb: ParamBuilder,
+        table_alias: str,
+        cast: typing.Optional[tsi_query.CastTo] = None,
+    ) -> str:
+        inner = super().as_sql(pb, table_alias)
+        return clickhouse_cast(f"{self.agg_fn}({inner})")
+
+
+class CallsMergedDynamicField(CallsMergedAggField):
+    extra_path: typing.Optional[list[str]] = None
+
+    def as_sql(
+        self,
+        pb: ParamBuilder,
+        table_alias: str,
+        cast: typing.Optional[tsi_query.CastTo] = None,
+    ) -> str:
+        res = super().as_sql(pb, table_alias)
+        if cast != "exists":
+            path_str = "'$'"
+            if self.extra_path:
+                param_name = pb.add_param(quote_json_path_parts(self.extra_path))
+                path_str = _param_slot(param_name, "String")
+            val = f"JSON_VALUE({res}, {path_str})"
+            return clickhouse_cast(val, cast)
+        else:
+            # UGG - this is an ugly part of clickhouse! It is really difficult to differentiate
+            # between null, non-existent, "", and "null"!
+            path_parts = []
+            if self.extra_path:
+                for part in self.extra_path:
+                    path_parts.append(", " + _param_slot(pb.add_param(part), "String"))
+            safe_path = "".join(path_parts)
+            return f"(NOT (JSONType({res}{safe_path}) = 'Null' OR JSONType({res}{safe_path}) IS NULL))"
+
+    def as_select_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        if self.extra_path:
+            raise NotImplementedError(
+                "Dynamic fields cannot be selected directly, yet - implement me!"
+            )
+        return f"{super().as_sql(pb, table_alias)} AS {self.field}"
+
+    def with_path(self, path: list[str]) -> "CallsMergedDynamicField":
+        extra_path = [*(self.extra_path or [])]
+        extra_path.extend(path)
+        return CallsMergedDynamicField(
+            field=self.field, agg_fn=self.agg_fn, extra_path=extra_path
+        )
+
+    def is_heavy(self) -> bool:
+        return True
+
+
+class OrderField(BaseModel):
+    field: CallsMergedField
+    direction: typing.Literal["ASC", "DESC"]
+
+    def as_sql(self, pb: ParamBuilder, table_alias: str) -> str:
+        options: list[typing.Tuple[typing.Optional[tsi_query.CastTo], str]]
+        if isinstance(self.field, CallsMergedDynamicField):
+            # Prioritize existence, then cast to float, then str
```

Comment:
Might be clearer to say double than float

Reply:
done